File: fddm-fpt/python/bench.py
"""full_ddm cost and accuracy vs hddm-wfpt (HSSM's blackbox likelihood) over parameter sets
drawn from the priors of Tran et al. (2021); see ../R/bench_sets.R.
  OMP_NUM_THREADS=1 python bench.py           ->  bench_py.csv
  SWEEP=1 OMP_NUM_THREADS=1 python bench.py   ->  curve_py.csv  (cost vs trials per evaluation)"""
import csv
import os
import sys
import time
import numpy as np
import ddmsz
from hddm_wfpt import wfpt

NRT = int(os.environ.get("NRT", 1000))
rng = np.random.default_rng(1)

g = np.genfromtxt("../R/bench_sets.csv", delimiter=",", names=True)   # Tran et al. priors
NSET = len(g)
sets = [dict(v=r["v"], a=r["a"]/2, z=r["w"], sz=r["sw"], sv=r["sv"],  # HSSM: a is half, t centred
             t=r["t0"] + r["st0"]/2, st=r["st0"]) for r in g]

def draw(n, P):
    """signed rts from the model itself, by rejection against the density's peak."""
    lo, hi = P["t"] - P["st"]/2 + 1e-4, P["t"] + P["st"]/2 + 4.0
    grid = np.linspace(lo, hi, 400)
    top = 1.05 * max(ddmsz.full_ddm(grid, np.full(400, h), **P)[0].max() for h in (0, 1))
    x = np.empty(0)
    while x.size < n:
        c, hit = rng.uniform(lo, hi, 8*n), rng.random(8*n) < 0.5
        keep = rng.uniform(0, top, c.size) < ddmsz.full_ddm(c, hit, **P)[0]
        x = np.concatenate([x, np.where(hit[keep], c[keep], -c[keep])])
    return x[:n]

def ours(P, tol):
    return lambda x: ddmsz.full_ddm(np.abs(x), x > 0, tol=tol, **P)[0]

def hddm(P, err):
    h = dict(v=P["v"], sv=P["sv"], a=2*P["a"], z=P["z"], sz=P["sz"], t=P["t"], st=P["st"])
    return lambda x: np.exp(wfpt.wiener_logp_array(
        x=x, err=err, **{k: np.full(x.size, float(v)) for k, v in h.items()}))

def timed(f, x, reps=10):
    ts = []
    for _ in range(reps):
        t0 = time.perf_counter()
        f(x)
        ts.append(time.perf_counter() - t0)
    return float(np.median(ts)) / x.size * 1e6

METHODS = [("ours 1e-12 (density + 7 grads)", ours, 1e-12),
           ("ours 1e-8", ours, 1e-8),
           ("hddm-wfpt err=1e-8", hddm, 1e-8),
           ("hddm-wfpt err=1e-4", hddm, 1e-4)]

if os.environ.get("SWEEP"):              # median over the first 5 sets, stop once a call passes 10 s
    rows = []
    for name, mk, prec in (METHODS[0], METHODS[2]):
        for n in (30, 100, 300, 1000, 3000, 10_000, 30_000, 100_000):
            ms = float(np.median([timed(mk(P, prec), draw(n, P), reps=3) * n / 1e3 for P in sets[:5]]))
            rows.append((name, n, ms))
            print(f"{name:30s} n={n:7d}  {ms:10.3f} ms", flush=True)
            if ms > 1e4:
                break
    with open("curve_py.csv", "w", newline="") as fh:
        csv.writer(fh).writerows([("method", "n", "ms")] + rows)
    sys.exit()

rows = []
for i, P in enumerate(sets, 1):
    x = draw(NRT, P)
    ref = ours(P, 1e-12)(x)
    big = ref > 1e-10
    for name, mk, prec in METHODS:
        f = mk(P, prec)
        us = timed(f, x)
        acc = 0.0 if name.endswith("grads)") else float(np.max(np.abs(f(x)[big]/ref[big] - 1)))
        rows.append((i, name, us, acc))
        print(f"set {i:2d}  {name:30s} {us:8.2f} us/trial  rel {acc:.1e}", flush=True)

with open("bench_py.csv", "w") as fh:
    fh.write("set,method,us_per_trial,rel_diff\n")
    for r in rows:
        fh.write("%d,%s,%.4f,%.3e\n" % r)

print(f"\n== median over {NSET} sets (IQR), accuracy vs ours at 1e-12 ==")
for name, _, _ in METHODS:
    u = np.array([r[2] for r in rows if r[1] == name])
    a = np.array([r[3] for r in rows if r[1] == name])
    q = np.percentile(u, [25, 75])
    print(f"{name:30s} {np.median(u):8.2f} us  [{q[0]:7.2f}, {q[1]:8.2f}]   "
          f"rel: median {np.median(a):.1e}  max {a.max():.1e}")

# anchor: the reference above is ours at 1e-12, so check it against mpmath at 30 digits
import mpmath as mp
sys.path.insert(0, "../../verify"); from highprec2 import g_eta
mp.mp.dps = 30

def hp(rt, P):                       # f = (1/st) int_u (1/sz) int_w g_eta(u, w) dw du
    v, a, z, t, sz, sv, st = (mp.mpf(P[k]) for k in ddmsz.PARAMS)
    g = lambda u, w: g_eta(u, v, sv, 2*a, w, J=60)
    gb = lambda u: mp.quad(lambda w: g(u, w), [z - sz/2, z + sz/2])/sz if sz > 0 else g(u, z)
    hi = mp.mpf(rt) - (t - st/2)
    lo = max(hi - st, mp.mpf(0))
    return mp.quad(gb, [lo, hi])/st if st > 0 else gb(hi)

print("\nmpmath anchor (30 digits), 3 rts on each of 3 sets:")
worst = 0.0
for P in sets[:3]:
    x = np.abs(draw(3, P))
    f = ddmsz.full_ddm(x, np.zeros(3), tol=1e-12, **P)[0]
    for xi, fi in zip(x, f):
        e = abs(fi/float(hp(xi, P)) - 1)
        worst = max(worst, e)
print(f"  ours at 1e-12 vs 30-digit reference: max rel {worst:.1e}")
assert worst < 1e-9, worst

File: fddm-fpt/R/bench.R
# Cost and accuracy vs rtdists and WienR, rtdists parametrization, over parameter sets drawn
# (EMC2's dDDM calls the same Hartmann-Klauer engine as WienR, at a default precision of 5e-3)
# from the priors of Tran et al. (2021), see bench_sets.R.  Response times are simulated from
# each set, so both barriers appear in the proportions the model implies.
#   OMP_NUM_THREADS=1 Rscript bench.R          ->  bench_r.csv
#   SWEEP=1 OMP_NUM_THREADS=1 Rscript bench.R  ->  curve_r.csv  (cost vs trials per evaluation)
suppressMessages({library(rtdists); library(WienR); library(microbenchmark)})
sink(tempfile()); source("ddmsz.R"); sink()

NRT <- as.integer(Sys.getenv("NRT", 1000))
set.seed(1)

g <- read.csv("bench_sets.csv")                 # Tran et al. priors, all seven parameters
NSET <- nrow(g)
sets <- lapply(seq_len(NSET), function(i) list(
  a = g$a[i], v = g$v[i], z = g$w[i] * g$a[i], sz = g$sw[i] * g$a[i], sv = g$sv[i],
  t0 = g$t0[i], st0 = g$st0[i]))

draw <- function(n, p) {                        # rts from the model, split by barrier
  d <- do.call(rdiffusion, c(list(n = n), p))
  split(d$rt, factor(d$response, c("lower", "upper")))
}
wpar <- function(p) list(a = p$a, v = p$v, w = p$z/p$a, t0 = p$t0, sv = p$sv,
                         sw = p$sz/p$a, st0 = p$st0)

# each method evaluates the whole data set: one call per barrier
byresp <- function(one) function(rts, p) unlist(lapply(names(rts), function(r)
  if (length(rts[[r]])) one(rts[[r]], r, p) else numeric(0)))

ours   <- byresp(function(rt, r, p) do.call(ddm_rtdists, c(list(rt, r), p))$density)
rtd    <- function(pr) byresp(function(rt, r, p)
  do.call(ddiffusion, c(list(rt, r, precision = pr), p)))
wienr  <- function(pr) byresp(function(rt, r, p)
  do.call(WienerPDF, c(list(rt, r), wpar(p), list(precision = pr, n.threads = FALSE)))$value)
wienrg <- function(pr) byresp(function(rt, r, p) {
  v <- do.call(WienerPDF, c(list(rt, r), wpar(p), list(precision = pr, n.threads = FALSE)))$value
  for (nm in c("a", "v", "w", "sv", "sw"))
    do.call(get(paste0("d", nm, "WienerPDF")),
            c(list(rt, r), wpar(p), list(precision = pr, n.threads = FALSE)))
  v
})

# name, function, trials per timing call, repetitions, response times used for accuracy
M <- list(
  list("ours 1e-12 (density + 7 grads)", ours,          NRT, 10, NRT),
  list("rtdists precision=3",            rtd(3),        NRT, 10, NRT),
  list("rtdists precision=5",            rtd(5),        100,  3, 100),
  list("rtdists precision=8",            rtd(8),         25,  2,  50),
  list("WienR 5e-3 (EMC2 default)",      wienr(5e-3),   NRT, 10, NRT),
  list("WienR 1e-4",                     wienr(1e-4),   NRT, 10, NRT),
  list("WienR 1e-8",                     wienr(1e-8),   NRT, 10, NRT),
  list("WienR 1e-12",                    wienr(1e-12),  200,  5, 200),
  list("WienR 1e-12 + 5 grads",          wienrg(1e-12), 200,  3, 200))

take <- function(rts, n) lapply(rts, function(x) head(x, max(1, round(n*length(x)/NRT))))

if (nzchar(Sys.getenv("SWEEP"))) {       # median over the first 5 sets, stop once a call passes 10 s
  sweep <- NULL
  for (m in M[-c(4, 6)]) {               # rtdists precision 8 and WienR 1e-4 add no new line shape
    for (n in c(30, 100, 300, 1000, 3000, 10000, 30000, 100000)) {
      ms <- median(sapply(sets[1:5], function(p) {
        rts <- draw(n, p)                 # simulate outside: microbenchmark re-evaluates its argument
        median(microbenchmark(m[[2]](rts, p), times = 3)$time) / 1e6
      }))
      sweep <- rbind(sweep, data.frame(method = m[[1]], n = n, ms = ms))
      cat(sprintf("%-30s n=%7d  %10.3f ms\n", m[[1]], n, ms)); flush.console()
      if (ms > 1e4) break
    }
  }
  write.csv(sweep, "curve_r.csv", row.names = FALSE)
  quit(save = "no")
}

out <- NULL
t_start <- Sys.time()
unlink("bench_r.csv")                              # rows are appended below, set by set
for (i in seq_along(sets)) {
  p <- sets[[i]]
  rts <- draw(NRT, p)
  for (m in M) {
    nm <- m[[1]]; f <- m[[2]]; reps <- m[[4]]
    r <- take(rts, min(m[[3]], NRT))
    nn <- sum(lengths(r))
    us <- median(microbenchmark(f(r, p), times = reps, unit = "us")$time)/1e3/nn
    k <- take(rts, min(m[[5]], NRT))
    acc <- if (identical(nm, M[[1]][[1]])) 0 else {
      v <- f(k, p); rf <- ours(k, p)          # compare on the same response times
      max(abs(v[rf > 1e-10]/rf[rf > 1e-10] - 1))
    }
    row <- data.frame(set = i, method = nm, us_per_trial = us, n = nn,
                      n_acc = sum(lengths(k)), rel_diff = acc)
    out <- rbind(out, row)
    write.table(row, "bench_r.csv", sep = ",", row.names = FALSE, append = TRUE,
                col.names = (i == 1 && identical(nm, M[[1]][[1]])))
    cat(sprintf("[%4.0f min] set %3d/%d  %-30s %10.2f us/trial  rel %.1e\n",
                as.numeric(difftime(Sys.time(), t_start, units = "mins")), i, length(sets), nm, us, acc))
    flush.console()
  }
}

cat("\n== median over", NSET, "sets (IQR), accuracy vs ours at 1e-12 ==\n")
for (m in M) {
  s <- out[out$method == m[[1]], ]
  u <- s$us_per_trial
  cat(sprintf("%-30s %10.2f us  [%8.2f, %9.2f]  %d sets   rel: median %.1e  max %.1e\n",
              m[[1]], median(u), quantile(u, .25), quantile(u, .75), nrow(s),
              median(s$rel_diff), max(s$rel_diff)))
}
cat(sprintf("\n%s | R %s | rtdists %s | WienR %s\n", Sys.info()[["sysname"]],
            getRversion(), packageVersion("rtdists"), packageVersion("WienR")))
